flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/StopWithSavepoint.java

```diff
@@ -134,30 +143,53 @@ public JobStatus getJobStatus() {
         return JobStatus.RUNNING;
     }
 
+    private void handleFailure(Failure failure) {
+        failureCollection.add(failure);
+        FailureResult failureResult = context.howToHandleFailure(failure);
+        transitionOnFailure(failureResult);
+    }
+
     @Override
     public void handleGlobalFailure(Throwable cause) {
-        handleAnyFailure(cause);
+        handleFailure(Failure.createGlobal(cause));
     }
 
     @Override
     boolean updateTaskExecutionState(TaskExecutionStateTransition taskExecutionStateTransition) {
         final boolean successfulUpdate =
                 getExecutionGraph().updateState(taskExecutionStateTransition);
 
-        if (successfulUpdate) {
-            if (taskExecutionStateTransition.getExecutionState() == ExecutionState.FAILED) {
-                Throwable cause = taskExecutionStateTransition.getError(userCodeClassLoader);
-                handleAnyFailure(
-                        cause == null
-                                ? new FlinkException(
-                                        "Unknown failure cause. Probably related to FLINK-21376.")
-                                : cause);
-            }
+        if (successfulUpdate
+                && taskExecutionStateTransition.getExecutionState() == ExecutionState.FAILED) {
+            handleFailure(
+                    Failure.createLocal(
+                            extractError(taskExecutionStateTransition),
+                            extractExecutionVertexID(taskExecutionStateTransition)));
         }
 
         return successfulUpdate;
     }
 
+    private void transitionOnFailure(FailureResult failureResult) {
```

Comment:
This is duplicated in `Executing` state.

Reply:
nvm, I've missed that `StateWithExecutionGraph` can't always transition to `restarting` and `failing`